### ISSUE TITLE
EFS attaching to controller nodes instead of workers

### ISSUE THREAD

**Issue by drywheat** [kubernetes-incubator/kube-aws#487]
kube-aws version **v0.9.5**

I just launched a new cluster this morning using the latest release of this tool and discovered something interesting. After providing a brand-new EFS instance ID in my cluster.yml, I noticed that `kube-aws` created files/mounts on my controller nodes and not my workers.

This seems like a bug to me.

**Comment by mumoshu** [comment]
Hi @drywheat, thanks for the report!
* Did it work for you kube-aws versions older than v0.9.5?
* Could you provide me outputs from `journalctl -u efs.service` and `systemctl status efs.service` on one of worker nodes, if any?

**Comment by drywheat** [comment]
Using **v0.9.3-rc.5**, I was able to attach and use EFS mounts on all of my worker nodes (also attached to controllers). Here's an example of the output from one of them:

```
core@old-worker ~ $ systemctl status efs
● efs.service
   Loaded: loaded (/etc/systemd/system/efs.service; disabled; vendor preset: disabled)
   Active: active (exited) since Thu 2017-03-30 22:00:34 UTC; 3 days ago
  Process: 1183 ExecStart=/bin/sh -c grep -qs /efs /proc/mounts || /usr/bin/mount -t nfs4 -o nfsvers=4.1,rsize=1048576,wsize=1048576,hard,
  Process: 1180 ExecStartPre=/usr/bin/mkdir -p /efs (code=exited, status=0/SUCCESS)
 Main PID: 1183 (code=exited, status=0/SUCCESS)
    Tasks: 0
   Memory: 0B
      CPU: 0
   CGroup: /system.slice/efs.service

Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.
```

Now, using kube-aws **v0.9.5**, the systemd unit file are NOT even created on the workers nodes:

```
core@new-worker ~ $ systemctl status efs
Unit efs.service could not be found.
core@new-worker ~ $ ls -l /etc/systemd/system/
total 40
-rw-r--r--. 1 root root  315 Apr  1 16:41 cfn-signal.service
drwxr-xr-x. 2 root root 4096 Apr  1 16:41 docker.service.d
drwxr-xr-x. 2 root root 4096 Apr  1 16:41 flanneld.service.d
drwxr-xr-x. 2 root root 4096 Apr  1 16:40 multi-user.target.requires
lrwxrwxrwx. 1 root root    9 Apr  1 16:40 user-configdrive.service -> /dev/null
lrwxrwxrwx. 1 root root    9 Apr  1 16:40 user-configvirtfs.service -> /dev/null
```

I've been trying to establish what changed by comparing the config files from both versions of the tool.  Without undertaking the painstaking of process of groking the entire CF JSON, I think I still might have a clue as to why this is happening.

**v0.9.3-rc.5** generates a single CF stack file (i.e. `stack-template.json`) and **v0.9.5** generates a couple of templates under the `stack-templates` directory. Closer inspection of those files shows that this was probably done to uncouple the control plan and worker configs.

The old CF stack template has a single EFS config that's applied to both the controller and worker.

```
sonny@workstation -> grep ElasticFileSystemID stack-template.json 
    {{if $.ElasticFileSystemID}}
    {{if $.ElasticFileSystemID}}
        "FileSystemId": "{{$.ElasticFileSystemID}}",
```

Whereas, I'm not seeing the same config in the `node-pool.json.tmpl`.

  ```
sonny@workstation -> ls -l stack-templates/
total 120
-rw-r--r--  1 sonny  staff  40063 Apr  3 10:04 control-plane.json.tmpl
-rw-r--r--  1 sonny  staff  13441 Apr  3 10:04 node-pool.json.tmpl
-rw-r--r--  1 sonny  staff   1409 Apr  3 10:04 root.json.tmpl
sonny@workstation -> grep -i ElasticFileSystemID stack-templates/*
stack-templates/control-plane.json.tmpl:    {{if $.ElasticFileSystemID}}
stack-templates/control-plane.json.tmpl:    {{if $.ElasticFileSystemID}}
stack-templates/control-plane.json.tmpl:        "FileSystemId": "{{$.ElasticFileSystemID}}",
```

**Comment by mumoshu** [comment]
Thanks for the detailed feedback and the study 👍 

To summarize:
* In stack templates, `.ElasticFileSystemId` is use only for creating `SecurityGroupMountTarget`
* As long as ElasticFileSystemId is provided from cluster.yaml to each node pool's cloud-config-worker, the `rpc-statd.service` is created and started to mount the EFS to the worker node.

So, it certainly seems to be a bug. Probably I broke it in v0.9.4 while migrating kube-aws to use cfn nested stacks? I'll investigate it further to fix asap!

**Comment by drywheat** [comment]
Thanks for provided further clarity.

I did some more testing this afternoon and now _believe_ this bug is caused by the `ElasticFileSystemID` not being populated in the node pool CF stack. Without that variable, neither the `rpc-statd.service` nor the `efs.service` get created on the worker nodes, which is what I was observing.

When I remove that conditional statement from the worker userdata file and perform and `kube-aws update`, the new workers have both services up and running properly.

**Comment by mumoshu** [comment]
Hi @drywheat, thanks for the additional info!
I had been thinking about this for a while.
How about leaving it to require you to explicitly specify `elasticFileSystemId` also for node pool(s) you want to mount the EFS?

So, a "proper" cluster.yaml after the decision would look like the below:

```yaml
elasticFileSystemId: efs-abcdefg
worker:
  nodePools:
  # If you'd like pool1 to mount the same efs as the control-plane
  - name: pool1
     elasticFileSystemId: efs-abcdefg
  # If you'd like pool2 to mount a different efs just for this node pool
  - name: pool2
     elasticFileSystemId: efs-hijklmn
```

The reason behind it is that, automatically propagating the `elasticFileSystemId` to every node pool ends the EFS to be unable to be mounted, when some of node pools were assigned to subnets other than the ones for the control-plane. An EFS is allowed to be mounted to only 1 subnet per AZ, by its design.

**Comment by mumoshu** [comment]
The key is explained in cluster.yaml as "This is a NFS share that will be available across the entire cluster", which is not true.
We should update cluster.yaml anyway.

**Comment by mumoshu** [comment]
Ah, just realized that our current implementation doesn't allow setting a different elasticFileSystemId only for a worker node pool like described in https://github.com/kubernetes-incubator/kube-aws/issues/487#issuecomment-292824475
Sorry for back and forth but we'd better make node pool's elasticFileSystemId to default to the top-level one for now.
If anyone likes to have a different set of EFS' per node pool, please submit a feature request dedicated to it!

**Comment by drywheat** [comment]
That makes sense to me as a patch that closely emulates the old behavior (pre-node-pool support).

**Comment by mumoshu** [comment]
@drywheat The fix is included in v0.9.6-rc.2 released today! Would you mind confirming if it works for you?

**Comment by drywheat** [comment]
Sure. I'll give it a shot tomorrow and see how it works.
